pac4j-kerberos/src/main/java/org/pac4j/kerberos/credentials/extractor/KerberosExtractor.java

```diff
@@ -0,0 +1,53 @@
+package org.pac4j.kerberos.credentials.extractor;
+
+import java.nio.charset.StandardCharsets;
+import java.util.Base64;
+
+import org.pac4j.kerberos.credentials.KerberosCredentials;
+import org.pac4j.core.context.HttpConstants;
+import org.pac4j.core.context.WebContext;
+import org.pac4j.core.credentials.extractor.CredentialsExtractor;
+import org.pac4j.core.exception.CredentialsException;
+import org.pac4j.core.exception.DeferredHttpAction;
+import org.pac4j.core.exception.DeferredHttpActionCallback;
+
+/**
+ * To extract Kerberos headers.
+ *
+ * @author Garry Boyce
+ * @since 1.9.1
+ */
+public class KerberosExtractor implements CredentialsExtractor<KerberosCredentials> {
+    private final String clientName;
+
+    public KerberosExtractor(final String clientName) {
+        this.clientName = clientName;
+    }
+
+    @Override
+    public KerberosCredentials extract(WebContext context) {
+        final String header = context.getRequestHeader(HttpConstants.AUTHORIZATION_HEADER);
+        if (header == null) {
+            throw DeferredHttpAction.deferredHttpAction("Kerberos Header not found", new DeferredHttpActionCallback() {

```

Comment:
Flexibility may happen at two level:
- at the clients level and this is true for any pac4j implementation. I have the feeling you need two clients: one direct and one indirect, both inheriting from `DirectClientV2` and `IndirectClientV2`and sharing the same components:  `KerberosExtractor`, `KerberosAuthenticator`... This is mostly where your efforts should be placed.
- at the filters level and most pac4j implementation should work like that. You can for example modify the `DefaultSecurityLogic` to take into account the fact you have only one direct client. But this is certainly a contribution you won't share as it meets only your needs.

You have to find the proper balancer between the generic Kerberos contribution and your specific needs.